### PR TITLE
SumT can now be melted, even inside ForeachFacts

### PR DESCRIPTION
**Pull request by jystic**

### CODE REVIEW

File: src/Icicle/Avalanche/Statement/Simp/Melt.hs [hunk 1]
```diff
@@ -21,35 +23,88 @@ import              P
...
+
+  , primMkOpt  :: ValType -> Name n -> Name n -> Exp a n p
+  , primIsSome :: ValType -> Exp a n p        -> Exp a n p
+  , primGet    :: ValType -> Exp a n p        -> Exp a n p
+
+  , primMkSum   :: ValType -> ValType -> Name n -> Name n -> Name n -> Exp a n p
+  , primIsRight :: ValType -> ValType -> Exp a n p                  -> Exp a n p
+  , primLeft    :: ValType -> ValType -> Exp a n p                  -> Exp a n p
+  , primRight   :: ValType -> ValType -> Exp a n p                  -> Exp a n p
+  }
```
Comment (jystic) [inline src/Icicle/Avalanche/Statement/Simp/Melt.hs]: I needed to use these in two functions, so I pulled it out to a record (designed to be unpacked with `RecordWildCards`, hope this is ok!)

Might be useful to be its own module as things are written longhand / duplicated in the constructors pass at the moment.

**Comment by tranma** [comment]
nice, was hoping to see this

File: src/Icicle/Avalanche/Statement/Simp/Melt.hs (hunk 1 above)
Comment (tranma) [inline src/Icicle/Avalanche/Statement/Simp/Melt.hs]: in Source/Core, there are `Combinators` modules where you can stick things like these. Might be worth adding one here

**Comment by tranma** [comment]
not sure if flatten_simp_commutes_value should still hold

**Comment by jystic** [comment]
^ yeah i'm fixing this at the moment, it's a good test so i don't want to break it

File: src/Icicle/Avalanche/Statement/Simp/Melt.hs (hunk 1 above)
Comment (jystic) [inline src/Icicle/Avalanche/Statement/Simp/Melt.hs]: I'm thinking I might move them to `Icicle.Avalanche.Combinators` in a separate PR?

**Comment by jystic** [comment]
Ok @tranma, `flatten_simp_commutes_value` should hold now

**Comment by tranma** [comment]
yup, fair enough :bomb:

**Pull request merged**